src/browser/api_protocol/api_handlers/window.js

```diff
@@ -81,474 +80,272 @@ module.exports.init = function() {
-
-    dataAck.data = Window.getBounds(windowIdentity);
-    ack(dataAck);
+function getWindowBounds(identity, message) {
+    return Window.getBounds(getWinId(message));
 }
 
-function focusWindow(identity, message, ack) {
-    var payload = message.payload,
-        windowIdentity = apiProtocolBase.getTargetWindowIdentity(payload);
-
-    Window.focus(windowIdentity);
-    ack(successAck);
+function focusWindow(identity, message) {
+    Window.focus(getWinId(message));
 }
 
-function flashWindow(identity, message, ack) {
-    var payload = message.payload,
-        windowIdentity = apiProtocolBase.getTargetWindowIdentity(payload);
-
-    Window.flash(windowIdentity);
-    ack(successAck);
+function flashWindow(identity, message) {
+    Window.flash(getWinId(message));
 }
 
-function enableWindowFrame(identity, message, ack) {
-    var payload = message.payload,
-        windowIdentity = apiProtocolBase.getTargetWindowIdentity(payload);
-
-    Window.enableFrame(windowIdentity);
-    ack(successAck);
+function enableWindowFrame(identity, message) {
+    Window.enableFrame(getWinId(message));
 }
 
 function executeJavascript(identity, message, ack, nack) {
-    let payload = message.payload;
-    let windowIdentity = apiProtocolBase.getTargetWindowIdentity(payload);
-    let pUuid = windowIdentity.uuid;
+    let pUuid = getWinId(message).uuid;
 
     while (pUuid) {
         if (pUuid === identity.uuid) {
-            return Window.executeJavascript(windowIdentity, payload.code, (err, result) => {
-                if (err) {
-                    nack(err);
-                } else {
-                    let dataAck = _.clone(successAck);
-                    dataAck.data = result;
-                    ack(dataAck);
-                }
+            return new Promise((resolve, reject) => {
+                Window.executeJavascript(getWinId(message), message.payload.code, (err, result) => {
+                    if (err) {
+                        reject(err);
+                    } else {
+                        resolve(result);
+                    }
+                });
             });
         }
         pUuid = Application.getParentApplication({
             uuid: pUuid
         });
-
     }
-    return nack(new Error('Rejected, target window is not owned by requesting identity'));
-}
 
-function disableWindowFrame(identity, message, ack) {
-    var payload = message.payload,
-        windowIdentity = apiProtocolBase.getTargetWindowIdentity(payload);
+    return Promise.reject(new Error('Rejected, target window is not owned by requesting identity'));
+}
 
-    Window.disableFrame(windowIdentity);
-    ack(successAck);
+function disableWindowFrame(identity, message) {
+    Window.disableFrame(getWinId(message));
 }
 
-function windowEmbedded(identity, message, ack) {
-    let payload = message.payload;
-    // Ensure expected shape for identity utility compliance
+function windowEmbedded(identity, message) {
```

Comment:
This needs to be manually tested, is not covered in the test apps. See @nicholasdgoodman